src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs

```diff
@@ -23,20 +28,43 @@ internal class VsCodeWindowManager : IVsCodeWindowManager, IVsCodeWindowEvents
             private readonly IVsCodeWindow _codeWindow;
             private readonly ComEventSink _sink;
             private readonly IOptionService _optionService;
+            private readonly IThreadingContext _threadingContext;
 
-            private INavigationBarController _navigationBarController;
-            private IVsDropdownBarClient _dropdownBarClient;
+            private INavigationBarController? _navigationBarController;
+            private IVsDropdownBarClient? _dropdownBarClient;
+            private WorkspaceRegistration _workspaceRegistration;
 
             public VsCodeWindowManager(TLanguageService languageService, IVsCodeWindow codeWindow)
             {
                 _languageService = languageService;
                 _codeWindow = codeWindow;
 
                 var workspace = languageService.Package.ComponentModel.GetService<VisualStudioWorkspace>();
-                _optionService = workspace.Services.GetService<IOptionService>();
+                _optionService = workspace.Services.GetRequiredService<IOptionService>();
+
+                _threadingContext = languageService.Package.ComponentModel.GetService<IThreadingContext>();
 
                 _sink = ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
                 _optionService.OptionChanged += OnOptionChanged;
+
+                _workspaceRegistration = GetWorkspaceRegistration();
+                _workspaceRegistration.WorkspaceChanged += OnWorkspaceRegistrationChanged;
+            }
+
+            private void OnWorkspaceRegistrationChanged(object sender, System.EventArgs e)
+            {
+                _threadingContext.JoinableTaskFactory.Run(async () =>
+                {
+                    // This event may not be triggered on the main thread, but adding and removing the navbar
+                    // must be done from the main thread.
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                    _navigationBarController?.SetWorkspace(_workspaceRegistration.Workspace);
```

Comment:
We probably should `#nullable enable` WorkspaceRegistration too since that might catch mistakes where we try to use WorkspaceRegistration.Workspace and don't realize it can be null.